libraries/rush-lib/src/logic/buildCache/ProjectBuildCache.ts

```diff
@@ -147,32 +147,69 @@ export class ProjectBuildCache {
 
     let localCacheEntryPath: string | undefined =
       await this._localBuildCacheProvider.tryGetCacheEntryPathByIdAsync(terminal, cacheId);
-    let cacheEntryBuffer: Buffer | undefined;
     let updateLocalCacheSuccess: boolean | undefined;
     if (!localCacheEntryPath && this._cloudBuildCacheProvider) {
       terminal.writeVerboseLine(
         'This project was not found in the local build cache. Querying the cloud build cache.'
       );
 
-      cacheEntryBuffer = await this._cloudBuildCacheProvider.tryGetCacheEntryBufferByIdAsync(
-        terminal,
-        cacheId
-      );
-      if (cacheEntryBuffer) {
-        try {
-          localCacheEntryPath = await this._localBuildCacheProvider.trySetCacheEntryBufferAsync(
-            terminal,
-            cacheId,
-            cacheEntryBuffer
-          );
-          updateLocalCacheSuccess = true;
-        } catch (e) {
-          updateLocalCacheSuccess = false;
+      if (this._cloudBuildCacheProvider.tryGetCacheEntryFileByIdAsync) {
+        const finalLocalCacheEntryPath: string = this._localBuildCacheProvider.getCacheEntryPath(cacheId);
+
+        // Derive the temp file from the destination path to ensure they are on the same volume
+        // In the case of a shared network drive containing the build cache, we also need to make
+        // sure the the temp path won't be shared by two parallel rush builds.
+        const randomSuffix: string = crypto.randomBytes(8).toString('hex');
+        const tempLocalCacheEntryPath: string = `${finalLocalCacheEntryPath}-${randomSuffix}.temp`;
```

Comment:
Do we do this when we're writing the buffer to disk? Seems like this code could be deduped a bit.